src/GafferSceneUI/ContextAlgo.cpp

```diff
@@ -73,6 +111,88 @@ PathMatcher getExpandedPaths( const Gaffer::Context *context )
 	return PathMatcher();
 }
 
+void expand( Context *context, const PathMatcher &paths, bool expandAncestors )
+{
+	PathMatcherData *expandedPaths = const_cast<PathMatcherData *>( context->get<PathMatcherData>( g_expandedPathsName, NULL ) );
+	if( !expandedPaths )
+	{
+		expandedPaths = new PathMatcherData();
+		context->set( g_expandedPathsName, expandedPaths );
+	}
+
+	PathMatcher &expanded = expandedPaths->writable();
+
+	bool needUpdate = false;
+	if( expandAncestors )
+	{
+		for( PathMatcher::RawIterator it = paths.begin(), eIt = paths.end(); it != eIt; ++it )
+		{
+			needUpdate |= expanded.addPath( *it );
+		}
+	}
+	else
+	{
+		for( PathMatcher::Iterator it = paths.begin(), eIt = paths.end(); it != eIt; ++it )
+		{
+			needUpdate |= expanded.addPath( *it );
+		}
+	}
+
+	if( needUpdate )
+	{
+		// We modified the expanded paths in place to avoid unecessary copying,
+		// so the context doesn't know they've changed. So we emit the changed
+		// signal ourselves
+		context->changedSignal()( context, g_expandedPathsName );
+	}
+}
+
+PathMatcher expandDescendants( Context *context, const PathMatcher &paths, const ScenePlug *scene, int depth )
+{
+	Context::Scope scope( context );
+
+	PathMatcherData *expandedPaths = const_cast<PathMatcherData *>( context->get<PathMatcherData>( g_expandedPathsName, NULL ) );
+	if( !expandedPaths )
+	{
+		expandedPaths = new PathMatcherData();
+		context->set( g_expandedPathsName, expandedPaths );
+	}
+
+	PathMatcher &expanded = expandedPaths->writable();
+
+	bool needUpdate = false;
+	PathMatcher leafPaths;
+
+	// \todo: parallelize the walk
+	for( PathMatcher::Iterator it = paths.begin(), eIt = paths.end(); it != eIt; ++it )
+	{
+		needUpdate |= expandWalk( *it, scene, depth + 1, expanded, leafPaths );
+	}
+
+	if( needUpdate )
+	{
+		// We modified the expanded paths in place to avoid unecessary copying,
+		// so the context doesn't know they've changed. So we emit the changed
+		// signal ourselves
+		context->changedSignal()( context, g_expandedPathsName );
+	}
+
+	return leafPaths;
+}
+
+void clearExpansion( Gaffer::Context *context )
+{
+	if( PathMatcherData *expandedPaths = const_cast<PathMatcherData *>( context->get<PathMatcherData>( g_expandedPathsName, NULL ) ) )
+	{
+		expandedPaths->writable().clear();
+
+		// We modified the expanded paths in place to avoid unecessary copying,
```

Comment:
PathMatcher copying is much cheaper than it was when I originally wrote that code, (it's just a recounted pointer assignment now), but modifying the copy will still have some overhead because we need to do a lazy-copy-on-write for any internal nodes that are edited. So the modify-in-place that `expandDescendants` is doing may still have an edge, so I think we should leave it as it is for now at least...